pom.xml

```diff
@@ -79,6 +79,15 @@
 
 	<dependencyManagement>
 		<dependencies>
+
+			<dependency>
```

Comment:
Can we add a comment as to why this is there? Also, if the application imports both, Spring Cloud GCP BOM and Spring Cloud BOM, would this still work? I think the version specified in Spring Cloud BOM will take precedence if the BOM is imported first.

Reply:
What about my question about the BOMs order?

Reply:
Missed the question, sorry. In Maven BOM import, the first version encountered wins. That's why I put `spring-cloud-function-dependencies` first. From [maven docs](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#transitive-dependencies): "Note that if two dependency versions are at the same depth in the dependency tree, the first declaration wins."

Do you mean if a customer puts Spring Cloud BOM in their application? The value of Spring Cloud GCP upgrading and releasing is for the customers who use our BOM to manage Spring Boot dependencies. If a customer application puts a different BOM first, there is nothing we can do about it.

I've actually realized that putting the version overrides in the Spring Cloud GCP parent is insufficient -- we have to add it to the `spring-cloud-gcp-dependencies` BOM, too. Otherwise the closest version ends up the one derived from spring-cloud-stream.

Reply:
Yes, I've tested one of the samples, and if the user puts Spring Cloud BOM above Spring Cloud GCP BOM, they will get the old version of functions:

```
    <dependencyManagement>

        <dependencies>

            <dependency>
                <groupId>org.springframework.cloud</groupId>
                <artifactId>spring-cloud-dependencies</artifactId>
                <version>2021.0.1</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>

            <dependency>
                <groupId>com.google.cloud</groupId>
                <artifactId>spring-cloud-gcp-dependencies</artifactId>
                <version>${project.version}</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
        </dependencies>
    </dependencyManagement>
```
 Running `mvn dependency:tree -Dincludes=org.springframework.cloud:spring-cloud-function-context`:

```
[INFO] --- maven-dependency-plugin:3.2.0:tree (default-cli) @ spring-cloud-gcp-pubsub-stream-sample ---
[INFO] com.google.cloud:spring-cloud-gcp-pubsub-stream-sample:jar:3.3.0-SNAPSHOT
[INFO] \- com.google.cloud:spring-cloud-gcp-pubsub-stream-binder:jar:3.3.0-SNAPSHOT:compile
[INFO]    \- org.springframework.cloud:spring-cloud-stream:jar:3.2.2:compile
[INFO]       \- org.springframework.cloud:spring-cloud-function-context:jar:3.2.2:compile

```

But if you have Spring Cloud GCP BOM first, the behavior is correct:

```
    <dependencyManagement>

        <dependencies>

            <dependency>
                <groupId>com.google.cloud</groupId>
                <artifactId>spring-cloud-gcp-dependencies</artifactId>
                <version>${project.version}</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>

            <dependency>
                <groupId>org.springframework.cloud</groupId>
                <artifactId>spring-cloud-dependencies</artifactId>
                <version>2021.0.1</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>

        </dependencies>
    </dependencyManagement>
```

```
[INFO] --- maven-dependency-plugin:3.2.0:tree (default-cli) @ spring-cloud-gcp-pubsub-stream-sample ---
[INFO] com.google.cloud:spring-cloud-gcp-pubsub-stream-sample:jar:3.3.0-SNAPSHOT
[INFO] \- com.google.cloud:spring-cloud-gcp-pubsub-stream-binder:jar:3.3.0-SNAPSHOT:compile
[INFO]    \- org.springframework.cloud:spring-cloud-stream:jar:3.2.2:compile
[INFO]       \- org.springframework.cloud:spring-cloud-function-context:jar:3.2.3:compile

```

Reply:
Unfortunately, it's common to have the Spring Cloud BOM first. For instance, the Spring Initializr does that by default. So, I'm not sure if we should add this temp fix as it may give folks a false sense of security.
Instead, it might be better to advise them to add the spring-cloud-function-dependencies directly in their application pom.xml.